mmv1/third_party/terraform/services/kms/data_source_google_kms_crypto_keys.go

```diff
@@ -0,0 +1,178 @@
+					* "name=projects/my-project/locations/global/keyRings/my-key-ring/cryptoKeys/my-key-1" will only retrieve a key with that exact name.
+					
+					[See the documentation about using filters](https://cloud.google.com/kms/docs/sorting-and-filtering)
+				`,
+			},
+			"keys": {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "A list of all the retrieved keys from the provided key ring",
+				Elem: &schema.Resource{
+					Schema: dsSchema,
+				},
+			},
+		},
+	}
+}
+
+func dataSourceGoogleKmsCryptoKeysRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*transport_tpg.Config)
+
+	keyRingId, err := parseKmsKeyRingId(d.Get("key_ring").(string), config)
+	if err != nil {
+		return err
+	}
+
+	id := fmt.Sprintf("%s/cryptoKeys", keyRingId.KeyRingId())
+	d.SetId(id)
+
+	log.Printf("[DEBUG] Searching for keys in key ring %s", keyRingId.KeyRingId())
+	res, err := dataSourceKMSCryptoKeysList(d, meta, keyRingId.KeyRingId())
+	if err != nil {
+		return err
+	}
+
+	// Check for cryptoKeys field, as empty response lacks keys
+	// If found, set data in the data source's `keys` field
+	if keys, ok := res["cryptoKeys"].([]interface{}); ok {
+		log.Printf("[DEBUG] Found %d keys in key ring %s", len(keys), keyRingId.KeyRingId())
+		value, err := flattenKMSKeysList(d, config, keys, keyRingId.KeyRingId())
+		if err != nil {
+			return fmt.Errorf("error flattening keys list: %s", err)
+		}
+		if err := d.Set("keys", value); err != nil {
+			return fmt.Errorf("error setting keys: %s", err)
+		}
+	} else {
+		log.Printf("[DEBUG] Found 0 keys in key ring %s", keyRingId.KeyRingId())
+	}
+
+	return nil
+}
+
+func dataSourceKMSCryptoKeysList(d *schema.ResourceData, meta interface{}, keyRingId string) (map[string]interface{}, error) {
+	config := meta.(*transport_tpg.Config)
+	userAgent, err := tpgresource.GenerateUserAgentString(d, config.UserAgent)
+	if err != nil {
+		return nil, err
+	}
+
+	url, err := tpgresource.ReplaceVars(d, config, "{{KMSBasePath}}{{key_ring}}/cryptoKeys")
+	if err != nil {
+		return nil, err
+	}
+
+	if filter, ok := d.GetOk("filter"); ok {
+		log.Printf("[DEBUG] Search for keys in key ring %s is using filter ?filter=%s", keyRingId, filter.(string))
+		url, err = transport_tpg.AddQueryParams(url, map[string]string{"filter": filter.(string)})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	billingProject := ""
+
+	if parts := regexp.MustCompile(`projects\/([^\/]+)\/`).FindStringSubmatch(url); parts != nil {
+		billingProject = parts[1]
+	}
+
+	// err == nil indicates that the billing_project value was found
+	if bp, err := tpgresource.GetBillingProject(d, config); err == nil {
+		billingProject = bp
+	}
+
+	headers := make(http.Header)
+	res, err := transport_tpg.SendRequest(transport_tpg.SendRequestOptions{
```

Comment:
This all looks good, my only concern is that it's not using `pageToken` to handle large lists of crypto keys. I can understand why this wasn't added though since it handles a large amount before needing to use the `nextPageToken` param. Regardless it would be worth adding support for `pageToken` to prevent any future panics for users.

Example of it being used:
https://github.com/GoogleCloudPlatform/magic-modules/blob/7381c52e624297d4328b71d0a0948f44f9d87dec/mmv1/third_party/terraform/services/storage/data_source_google_storage_buckets.go#L68-L104

Reply:
I'm unsure about this now since I ran into a quota limit when implementing on `key_rings`

https://github.com/GoogleCloudPlatform/magic-modules/blob/7381c52e624297d4328b71d0a0948f44f9d87dec/mmv1/third_party/terraform/services/secretmanager/data_source_secret_manager_secrets.go#L122-L129

Reply:
PR is ready to be merged. I would want to know whether it's a good idea to include pageToken support or not. Their would be some work needed on the tests in order to support it. If we're fine with not adding support i can approve this PR.